src/metrics.ts

```diff
@@ -0,0 +1,140 @@
+import { PrometheusMetrics, Bridge } from "matrix-appservice-bridge";
+import { Gauge, Counter, Histogram } from "prom-client";
+import { Log } from "./log";
+
+const AgeCounters = PrometheusMetrics.AgeCounters;
+const log = new Log("BridgeMetrics");
+const REQUEST_EXPIRE_TIME_MS = 30000;
+
+interface IAgeCounter {
+    setGauge(gauge: Gauge, morelabels: string[]);
+    bump(age: number);
+}
+
+interface IBridgeGauges {
+    matrixRoomConfigs: number;
+    remoteRoomConfigs: number;
+    matrixGhosts: number;
+    remoteGhosts: number;
+    matrixRoomsByAge: IAgeCounter;
+    remoteRoomsByAge: IAgeCounter;
+    matrixUsersByAge: IAgeCounter;
+    remoteUsersByAge: IAgeCounter;
+}
+
+export interface IBridgeMetrics {
+    registerRequest(id: string);
+    requestOutcome(id: string, isRemote: boolean, outcome: string);
+    remoteCall(method: string);
+    setPresenceCount(count: number);
+    storeCall(method: string, cached: boolean);
+}
+
+export class DummyBridgeMetrics implements IBridgeMetrics {
+    public registerRequest() {}
+    public requestOutcome() {}
+    public remoteCall() {}
+    public setPresenceCount() {}
+    public storeCall() {}
+}
+
+export class MetricPeg {
+    public static get get(): IBridgeMetrics {
+        return this.metrics;
+    }
+
+    public static set(metrics: IBridgeMetrics) {
+        this.metrics = metrics;
+    }
+
+    private static metrics: IBridgeMetrics = new DummyBridgeMetrics();
+}
+
+export class PrometheusBridgeMetrics implements IBridgeMetrics {
+    private metrics;
+    private remoteCallCounter: Counter;
+    private storeCallCounter: Counter;
+    private presenceGauge: Gauge;
+    private remoteRequest: Histogram;
+    private matrixRequest: Histogram;
+    private requestsInFlight: Map<string, number>;
+    private bridgeGauges: IBridgeGauges = {
+        matrixGhosts: 0,
```

Comment:
are these autofilled somehow?

Reply:
Yes, by the bridge lib. Though some of them will remain 0